lib/fluxsink/a2rfluxsink.cc

```diff
@@ -0,0 +1,187 @@
+#include "globals.h"
+#include "flags.h"
+#include "fluxmap.h"
+#include "bytes.h"
+#include "protocol.h"
+#include "fluxsink/fluxsink.h"
+#include "decoders/fluxmapreader.h"
+#include "lib/fluxsink/fluxsink.pb.h"
+#include "proto.h"
+#include "fmt/format.h"
+#include "fluxmap.h"
+#include "a2r.h"
+#include <fstream>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+namespace
+{
+uint32_t ticks_to_a2r(uint32_t ticks) {
+    return ticks * NS_PER_TICK / A2R_NS_PER_TICK;
+}
+
+bool singlesided(void) {
+	return config.heads().start() == config.heads().end();
+}
+
+class A2RFluxSink : public FluxSink
+{
+public:
+	A2RFluxSink(const A2RFluxSinkProto& lconfig):
+		_config(lconfig),
+		_bytes{},
+		_writer{_bytes.writer()}
+	{
+
+		std::cout << fmt::format("A2R: writing A2R {} file containing {} tracks\n",
+			singlesided() ? "single sided" : "double sided",
+			config.cylinders().end() - config.cylinders().start() + 1
+		);
+
+		char image_date[100];
+		time_t now{std::time(nullptr)};
+		strftime(image_date, sizeof(image_date), "%FT%TZ", gmtime(&now));
+		_metadata["image_date"] = image_date;
+
+		writeHeader();
+		writeInfo();
+		writeMeta();
+	}
+
+	~A2RFluxSink()
+	{
+		std::cout << "A2R: writing output file...\n";
+		std::ofstream of(_config.filename(), std::ios::out | std::ios::binary);
+		if (!of.is_open())
+			Error() << "cannot open output file";
+		_bytes.writeTo(of);
+		of.close();
+	}
+
+private:
+	void writeData(const auto data, size_t size) {
+		_writer += Bytes(reinterpret_cast<const uint8_t*>(data), size);
+	}
+	void writeChunkAndData(uint32_t chunk_id, const auto data, size_t size) {
+		_writer.write_le32(chunk_id);
+		_writer.write_le32(size);
+		writeData(data, size);
+	}
+
+	void writeHeader() {
+		writeData(a2r2_fileheader, sizeof(a2r2_fileheader));		  
+	}
+
+	void writeInfo() {
+		A2RInfoData info{
+		    A2R_INFO_CHUNK_VERSION,
+		    {}, // to be filled
+		    singlesided() ? A2R_DISK_525 : A2R_DISK_35,
+		    1,
+		    1,
+		};
+		char buf[33];
+		snprintf(buf, sizeof(buf), "Fluxengine %d%32.32s", FLUXENGINE_VERSION, "");
```

Comment:
fmt::format() here instead of printf, please!

Reply:
OK. is there a 'right' way to get the fluxengine software version? I think FLUXENGINE_VERSION is the device/protocol version? Or it can be left off.

Reply:
Oops, missed that. FluxEngine doesn't really have a version, I'm afraid. FLUXENGINE_VERSION is, as you say, the protocol version and isn't related to the client.